src/main/java/network/brightspots/rcv/ResultsWriter.java

```diff
@@ -136,18 +137,27 @@ private static String generateCvrSnapshotId(String cvrId, Integer round) {
         : String.format("ballot-%s", cvrId);
   }
 
-  private static String getCdfIdForCandidateCode(String code) {
-    String id = candidateCodeToCdfId.get(code);
+  // generates an internal ContestSelectionId based on a candidate code
+  private static String getContestSelectionIdForCandidateCode(String code) {
```

Comment:
Yeah sorry about that @tarheel!  I did not give you much explanation.

Among the important things:
1) All static election objects (Candidates, Contests, ContestIds) are parsed at beginning of cvr processing.  This simplifies and clarifies the parsing somewhat.
2) **CVRContestSelections now contain links to Candidate objects.**  During cvr processing these links are used to look up the linked Candidate.  From there, Candidate Name field is used to assign the rankings and create the CastVoteRecord for tabulation.  Previously, Candidate names were embedded directly in the CVRContestSelection objects using a "Code" object with a 'vendor-label' field on it.  This is non-standard, and while it's conceivable a vendor might do this (Code objects allow flexibility) it's not the primary way CDF encodes information about a Candidate.  The primary way is by linking to a Candidate object.
3) **Candidate names are stored in config** - not pulled from CDF file at tabulation time.  This normalizes candidate logic across formats.
4) **Configs do not use CDF internal ids to as candidate codes.**  It's important to understand that CDF internal object links (e.g. CandidateIds contained in a CVRContestSelection in CDF should be used only _within_ the context of the CDF.  They are generated by the cvr report generating program, and should not appear in our config files.
5) **Filter contest tabulation by contest ID as** specified in the config file

So to your second question: yes it would be possible for these methods to share an interface.  I think the real move is for all reader objects to share an interface.  I don't think what you propose would be difficult, nevertheless it's outside the scope of this PR.

Reply:
Could you create a follow-up issue (not for this deadline) to refactor this class?

Reply:
Sure, we already have such a ticket: https://github.com/BrightSpots/rcv/issues/251